Upgrade to Eclipse lsp4j

darora opened the pull request:
Fixes #150

Commit:
Initial upgrade to Eclipse lsp4j libs

Commit:
Sprinkle some assertj

Commit:
Minor stuff

Commit:
cs/findbugs

Commit:
Remove BS

jiahuijiang commented on groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServer.java:
```diff
@@ -16,56 +16,52 @@
 
     public GroovyLanguageServer(LanguageServerState state, TextDocumentService textDocumentService,
-            WorkspaceService workspaceService, WindowService windowService) {
+            WorkspaceService workspaceService) {
```
nitpick: all on the 2nd line

jiahuijiang commented on groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServer.java:
```diff
@@ -148,11 +131,15 @@ public static void main(String[] args) {
         LanguageServerState state = new DefaultLanguageServerState();
         LanguageServer server =
                 new GroovyLanguageServer(state, new DefaultTextDocumentService(state),
```
nitpick:
LanguageServer server = new GroovyLanguageServer(
    state, new DefaultTextDocumentService(state), new DefaultWorkspaceService(state));

jiahuijiang commented on groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServer.java:
```diff
@@ -148,11 +131,15 @@ public static void main(String[] args) {
+    public void connect(LanguageClient client) {
+        state.setPublishDiagnostics(client::publishDiagnostics);
+        state.setTelemetryEvent(client::telemetryEvent);
+    }
```
also set showMessage, showMessageRequest and logMessage

jiahuijiang commented on groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyWorkspaceCompiler.java:
```diff
@@ -241,13 +240,22 @@ private void resetCompilationUnit() {
-                    (value) -> new PublishDiagnosticsParamsBuilder().uri(workspaceRoot.toUri().toString()))
-                    .diagnostic(new DefaultDiagnosticBuilder(message.getMessage(), DiagnosticSeverity.Warning).build());
+            String message1 = message.getMessage();
+            checkNotNull(message1, "message cannot be null");
```
if it's null, convert it into empty? because groovy's warningMessage;s message is nullable

jiahuijiang commented on groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyWorkspaceCompiler.java:
```diff
@@ -241,13 +240,22 @@ private void resetCompilationUnit() {
+            checkNotNull(message1, "message cannot be null");
+            Diagnostic diag = new Diagnostic(Ranges.UNDEFINED_RANGE, message1, DiagnosticSeverity.Warning,
+                    GroovyConstants.GROOVY_COMPILER);
+            diagnosticsByFile.compute(workspaceRoot.toUri(),
```
diagnosticsByFile.computeIfAbsent(workspaceRoot.toUri(), (key) -> List.newArrayList())
    .add(diaf)?

jiahuijiang commented on groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyWorkspaceCompiler.java:
```diff
@@ -261,20 +269,29 @@ private void resetCompilationUnit() {
                 uri = workspaceRoot.toUri();
-                diagnostic = new DefaultDiagnosticBuilder(data.toString(), DiagnosticSeverity.Error).build();
+                String message1 = data.toString();
+                checkNotNull(message1, "message cannot be null");
```
same convert message

darora replied:
hmm I don't think the result of the `toString()` would be null?

jiahuijiang replied:
ohhh oops, nvm thought this is the same as the above

jiahuijiang commented on groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyWorkspaceCompiler.java:
```diff
@@ -261,20 +269,29 @@ private void resetCompilationUnit() {
             }
-            diagnosticsByFile.computeIfAbsent(uri, (value) -> new PublishDiagnosticsParamsBuilder().uri(uri.toString()))
-                    .diagnostic(diagnostic);
+            diagnosticsByFile.compute(uri, (key, existingValue) -> {
```
same computeIfAbsent to add empty list, then add "diagnostic"

jiahuijiang commented on groovy-language-server/src/main/java/com/palantir/ls/groovy/MethodVisitor.java:
```diff
@@ -219,33 +218,34 @@ private Location createLocation(URI locationUri, ASTNode node) {
     private SymbolInformation getVariableSymbolInformation(Variable variable) {
-        SymbolInformationBuilder builder =
-                new SymbolInformationBuilder().name(variable.getName());
+        String containerName;
```
make these final or create a "new SymbolInformation()" first then set each property below

jiahuijiang commented on groovy-language-server/src/test/java/com/palantir/ls/groovy/GroovyLanguageServerIntegrationTest.java:
```diff
@@ -86,81 +83,103 @@
+        // Start Groovy language server
+        createAndLaunchLanguageServer(serverInputStream, serverOutputStream);
+        int counter = 0;
+        while (server != null && counter++ < 20) {
```
when will this server be set to null?

darora replied:
if the other thread hasn't been scheduled for long enough

jiahuijiang commented on groovy-language-server/src/test/java/com/palantir/ls/groovy/GroovyWorkspaceCompilerTest.java:
```diff
@@ -176,13 +175,13 @@ public void testCompile_WithJavaExtensionError() throws InterruptedException, Ex
+        assertEquals(Sets.newHashSet(new PublishDiagnosticsParams(
+                        test.toPath().toUri().toString(),
+                        ImmutableList.of(
+                                new Diagnostic(Ranges.createRange(0, 19, 0, 26),
```
newline for the 1st arg

JiahuiJiang commented on language-server-commons/src/main/java/com/palantir/ls/services/AbstractTextDocumentService.java:
```diff
@@ -172,9 +171,13 @@ final void assertFileExists(URI uri) {
-    public CompletableFuture<CompletionList> completion(TextDocumentPositionParams position) {
-        return CompletableFuture.completedFuture(getState().getCompilerWrapper().getCompletion(
-                Uris.resolveToRoot(getWorkspacePath(), position.getTextDocument().getUri()), position.getPosition()));
+    public CompletableFuture<Either<List<CompletionItem>, CompletionList>> completion(
```
woooooooow

jiahuijiang commented on language-server-commons/src/test/java/com/palantir/ls/services/DefaultTextDocumentServiceTest.java:
```diff
@@ -104,65 +99,48 @@ public void setup() throws IOException {
+        CompletionItem methodA = new CompletionItem("methodA");
+        methodA.setKind(CompletionItemKind.Method);
+        expectedCompletionList = new CompletionList(
+                false,
```
nitpick: same line?

jiahuijiang commented on groovy-language-server/src/test/java/com/palantir/ls/groovy/GroovyLanguageServerIntegrationTest.java:
```diff
@@ -86,81 +83,103 @@
+    private void createAndLaunchLanguageServer(final InputStream in, final OutputStream out) {
+        new Thread(() -> {
+            LanguageServerState state = new DefaultLanguageServerState();
+            actualServer = new GroovyLanguageServer(
```
make arguments on the same line for these two constructors? no strong opinion

jiahuijiang commented on language-server-commons/src/main/java/com/palantir/ls/util/LoggerMessageTracer.java:
```diff
@@ -1,72 +0,0 @@
-import java.io.StringWriter;
-import org.slf4j.Logger;
-
-public class LoggerMessageTracer implements MessageTracer {
```
I think we still want to implement this. Right now all the requests/responses will be logged in language server logs. But right now it only logs errors https://github.com/eclipse/lsp4j/blob/bd1a4ddf97ea3a0ee9031ade9b10410bfb093999/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/StreamMessageProducer.java#L109

We should implement streamMessageProducer and streamMessageConsumer again
https://github.com/eclipse/lsp4j/issues/23

jiahuijiang reviewed (changes requested):
We can remove
"https://oss.sonatype.org/content/repositories/snapshots/" from build.gradle - allprojects.repositories now

Commit:
CR

darora commented:
@JiahuiJiang think I addressed all your comments. Added a DelegatingOutputStream that hooks into the existing LSPLauncher methods, rather than implementing our own subclass of the Consumer/Producer

jiahuijiang commented on groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyTreeParser.java:
```diff
@@ -291,27 +290,27 @@ private static SymbolKind getKind(ClassNode node) {
     private SymbolInformation getVariableSymbolInformation(String parentName, URI sourceUri, Variable variable) {
-        SymbolInformationBuilder builder =
-                new SymbolInformationBuilder().name(variable.getName()).containerName(parentName);
+        SymbolKind kind;
```
final here too

jiahuijiang commented on language-server-commons/src/main/java/com/palantir/ls/services/AbstractTextDocumentService.java:
```diff
@@ -103,10 +102,10 @@ public final void didSave(DidSaveTextDocumentParams params) {
-    public final void onPublishDiagnostics(Consumer<PublishDiagnosticsParams> callback) {
-        getState().setPublishDiagnostics(callback);
-    }
+//    @Override
```
remove?

jiahuijiang commented on language-server-commons/src/main/java/com/palantir/ls/util/CompletionUtils.java:
```diff
@@ -17,28 +17,26 @@
-                    .label(symbol.getName())
-                    .kind(symbolKindToCompletionItemKind(symbol.getKind()))
-                    .build());
+            CompletionItem completionItem = new CompletionItem(symbol.getName());
```
symbols.stream.map(****).collect()

jiahuijiang commented on language-server-commons/src/main/java/com/palantir/ls/util/DelegatingOutputStream.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Override
+    public void close() throws IOException {
+        closed = true;
```
do we need to flush first before closing? or is it already taken cared of?

darora replied:
Not really something demanded by the OutputStream interface, but does seem to be more of a convention. Added.

jiahuijiang reviewed (approved):
small nit-picks and one question in delegateOutputStream

Commit:
CR

Commit:
Remove old code

Commit:
CS

Commit:
Suppress upstream bugs
